src/app/cmdoptions.cpp

```diff
@@ -46,6 +48,23 @@
 
 namespace
 {
+    QStringList readParamsFile()
+    {
+        QString paramsFile = QCoreApplication::applicationFilePath() + QLatin1String(".params");
+        if (QFileInfo::exists(paramsFile)) {
```

Comment:
Is it necessary to check before opening?

Reply:
Probably not, but I don't need the `QFile` object when the params file does not exists (i.e. most of the times).

Reply:
You can open and read it inside local scope.

Reply:
Sorry, I don't understand what do you suggest.

Reply:
Sorry, I didn't see all affected code...
>I don't need the QFile object when the params file does not exists (i.e. most of the times).

But you just return in this case!

Reply:
>Probably not, but I don't need the QFile object when the params file does not exists (i.e. most of the times).

I really don't know it is that unfavorable to have `QFile` unconditionally (might be more convincing if you elaborate a bit more)... at least please invert this `if ()` check to return early.

Reply:
>  But you just return in this case!

Sorry, still can't understand.

> it is that unfavorable to have QFile unconditionally (might be more convincing if you elaborate a bit more)

I don't know what can I add. "The `QFile` class provides an interface for reading from and writing to files", quoting Qt docs. A file does not exist, why would I need an object to access it? Additionally, existing but unreadable files is an error, I must inform user about that.

> please invert this if () check to return early.

But then I would need a second `return {}`.

Reply:
>existing but unreadable files is an error, ...

Ok. Since in this particular case "absence of file" and "corrupted/unreadable file" have really different meaning you need to handle them differently.
>I must inform user about that.

But debug message isn't for user. Also I'm not sure it will be printed by "release" binary.
Maybe print it to the log?

Reply:
The application object is not created yet when this code is running, how can I access the log?

Reply:
But qDebug is not an option. std::cerr?

Reply:
or much lightweight `fprintf(stderr, )`

Reply:
I would even interrupt the app in this case. The user can rely on some parameters so that he can get an unexpected effect if the application starts without them.

Reply:
This sounds reasonable. I'm going to replace qDebug with direct write to cerr and throw an exception after that. Does anybody object?

Reply:
>Does anybody object?

Seems no one, let's try getting this into v4.1.5